### PR TITLE
Fix to ensure endpoints distinguish between form and query parameters

### PR DESCRIPTION
**Pull request by gregecho**
Fix gh-1451, initial implementation of get parameters from body for OAuth2AuthorizationCodeAuthenticationConverter.

### CODE REVIEW

**Comment by gregecho** [comment]
@jgrandja , I just modify the OAuth2AuthorizationCodeAuthenticationConverter to use the method `getFormParameters` in this commit. I'd like to get your input as to the approach for this implementation, then I'll handle the other AuthenticationConverters. 
Let me know if you need further clarification on any part of this draft.

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/web/authentication/OAuth2EndpointUtils.java [hunk 1]
```diff
@@ -54,6 +57,49 @@ static MultiValueMap<String, String> getParameters(HttpServletRequest request) {
 		return parameters;
 	}
 
+	static MultiValueMap<String, String> getFormParameters(HttpServletRequest request) {
```
Comment (jgrandja) [inline oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/web/authentication/OAuth2EndpointUtils.java:60]: This method can be simplified. Given the existing `getParameters()` method, the only change required to make it function as `getFormParameters()` is:

```java
...

parameterMap.forEach((key, values) -> {
	if (!request.getQueryString().contains(key) &&	// If not query parameter then it's a form parameter
			values.length > 0) {
		for (String value : values) {
			parameters.add(key, value);
		}
	}
});

...
```

And for `getQueryParameters()`, applying `if (request.getQueryString().contains(key) ...` would work.

File: oauth2-authorization-server/src/test/java/org/springframework/security/oauth2/server/authorization/config/annotation/web/configurers/OAuth2AuthorizationCodeGrantTests.java [hunk 2]
```diff
@@ -355,7 +356,8 @@ private OAuth2AccessTokenResponse assertTokenRequestReturnsAccessTokenResponse(R
 			OAuth2Authorization authorization, String tokenEndpointUri) throws Exception {
 		MvcResult mvcResult = this.mvc.perform(post(tokenEndpointUri)
 				.params(getTokenRequestParameters(registeredClient, authorization))
-				.header(HttpHeaders.AUTHORIZATION, getAuthorizationHeader(registeredClient)))
+				.header(HttpHeaders.AUTHORIZATION, getAuthorizationHeader(registeredClient))
+						.contentType(MediaType.APPLICATION_FORM_URLENCODED_VALUE))
```
Comment (jgrandja) [inline oauth2-authorization-server/src/test/java/org/springframework/security/oauth2/server/authorization/config/annotation/web/configurers/OAuth2AuthorizationCodeGrantTests.java:360]: Please remove all changes where the content type is set in the request as it's redundant. Let's keep the changes in this PR to only what's required for the fix.

**Review by jgrandja — CHANGES_REQUESTED** [review]
Thanks for the PR @gregecho. Please see review comments.

File: oauth2-authorization-server/src/test/java/org/springframework/security/oauth2/server/authorization/config/annotation/web/configurers/OAuth2AuthorizationCodeGrantTests.java (hunk 2 above)
Comment (gregecho) [reply oauth2-authorization-server/src/test/java/org/springframework/security/oauth2/server/authorization/config/annotation/web/configurers/OAuth2AuthorizationCodeGrantTests.java]: Reverted all changes for the content type changes.

**Comment by gregecho** [comment]
@jgrandja ,

I just changed the `/oauth2/token` endpoints' related converter to distinguish between form and query parameters in this PR as the required in original [gh-1451](https://github.com/spring-projects/spring-authorization-server/issues/1451).

Do we need to create a new enhancement ticket for the other endpoints?

**Comment by jgrandja** [comment]
@gregecho 

> Do we need to create a new enhancement ticket for the other endpoints?

No. Please include the updates to all `AuthenticationConverter`'s. Also, please remove `OAuth2EndpointUtils.getParameters()` as it's not needed anymore.

Lastly, I noticed some formatting changes, e.g. reorganized imports. Please ensure the only code changes in this PR are related to the fix only. Thank you.

**Comment by gregecho** [comment]
> @gregecho
> 
> > Do we need to create a new enhancement ticket for the other endpoints?
> 
> No. Please include the updates to all `AuthenticationConverter`'s. Also, please remove `OAuth2EndpointUtils.getParameters()` as it's not needed anymore.
> 
> Lastly, I noticed some formatting changes, e.g. reorganized imports. Please ensure the only code changes in this PR are related to the fix only. Thank you.

Thanks for review the PR. I have updated the PR accordingly.

**Comment by jgrandja** [comment]
Thanks for the updates @gregecho.

FYI, I added a polish commit to get this merged before tomorrow's release.

**Pull request closed without merging**